engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/MetaBuiltinNode.java

```diff
@@ -0,0 +1,55 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.type.TypesGen;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "meta_builtin",
+    description = "Find the type and constructs a Meta instances.",
+    autoRegister = false)
+public class MetaBuiltinNode extends Node {
+  Object execute(long index, @AcceptsError Object value, Array factories) {
+    if (index < 0) {
+      index = findIndex(value);
+    }
+    if (index < 0) {
+      return -index;
+    }
+    var factory = (AtomConstructor) factories.getItems()[(int) index];
+    return factory.newInstance(value);
+  }
+
+  private int findIndex(Object value) {
+    if (TypesGen.isFunction(value)) {
+      return 0;
+    }
+    if (value instanceof Atom) {
+      return 1;
+    }
+    if (TypesGen.isAtomConstructor(value)) {
+      return 2;
+    }
+    if (TypesGen.isUnresolvedSymbol(value) || TypesGen.isUnresolvedConversion(value)) {
+      return 4;
+    }
+    if (TypesGen.isDataflowError(value)) {
+      return -5;
```

Comment:
These indices look rather magical, I think we should have some comment explaining where they come from.

Why is the dataflow index the only negative one?

Reply:
This is the commit to read: 133ba8da13658fd5f9c866865d1a0bebabfbd668

Reply:
I think it would be good to have some comment explaining this logic, as it's not that easy to understand it.

Reply:
What is the result of `Meta.meta (Error.throw "Foo")`? 

IIRC it used to be `Error.Value`, but how I understand the current code, I'd guess it to be `Primitive.Value` (for `Text`). Do I understand it correctly?